module/VuFind/src/VuFind/Db/AdapterFactory.php

```diff
@@ -233,26 +240,69 @@ public function getAdapterFromConnectionString(
 
-        $driverName = $this->getDriverName($type);
-        $driverOptions = $this->getDriverOptions($driverName);
-
-        // Set up default options:
-        $options = [
-            'driver' => $driverName,
+        return $this->getAdapterFromArray([
+            'driver' => $type,
             'hostname' => $host,
             'username' => $username,
             'password' => $password,
             'database' => $dbName,
             'use_ssl' => $this->config->Database->use_ssl ?? false,
-            'driver_options' => $driverOptions,
+            'port' => $port ?? null,
+        ]);
+    }
+
+    /**
+     * Obtain a Laminas\DB connection using the config.
+     *
+     * @param string $overrideUser Username override (leave null to use username from config)
+     * @param string $overridePass Password override (leave null to use password from config)
+     *
+     * @return Adapter
+     */
+    public function getAdapterFromConfig($overrideUser = null, $overridePass = null)
+    {
+        if (!isset($this->config->Database)) {
+            throw new \Exception('"database" settings missing');
+        }
+        $config = $this->config->Database;
+        return $this->getAdapterFromArray([
+            'driver' => $config->database_driver ?? null,
+            'hostname' => $config->database_host ?? null,
+            'username' => $overrideUser ?? $config->database_username ?? null,
+            'password' => $overridePass ?? $this->getSecretFromConfig($config, 'database_password'),
+            'database' => $config->database_name,
+            'port' => $config->database_port ?? null,
+        ]);
+    }
+
+    /**
+     * Obtain a Laminas\DB connection using a set of given element.
+     *
+     * @param array $config Config array to connect to the DB containing
+     * driver (ie: mysql), username, password, hostname, database (db name), port
+     *
+     * @return Adapter
+     */
+    public function getAdapterFromArray(array $config)
+    {
+        $driverName = $this->getDriverName($config['driver']);
+
+        // Set up default options:
+        $options = [
+            'driver' => $driverName,
+            'hostname' => $config['hostname'] ?? null,
+            'username' => $config['username'] ?? null,
+            'password' => $config['password'] ?? null,
+            'database' => $config['database'] ?? null,
+            'use_ssl' => $this->config->Database->use_ssl ?? false,
+            'driver_options' => $this->getDriverOptions($driverName),
         ];
-        if (!empty($port)) {
-            $options['port'] = $port;
+        if (isset($config['port'])) {
+            $options['port'] = $config['port'];
         }
         // Get extra custom options from config:
-        $extraOptions = isset($this->config->Database->extra_options)
-            ? $this->config->Database->extra_options->toArray()
-            : [];
+        $extraOptions = isset($this->config->Database->extra_options) ?
```

Comment:
I think you could further simplify this to `$extraOptions = $this->config?->Database?->extra_options?->toArray() ?? [];`